Added search by comma separated list feature

ningxie1991 opened the pull request:
For issue #198

1. Modified the AsyncSelect component by manually setting value, inputValue and menuIsOpen
2. Added handleInputChange, handleCommaSeparatedInput methods to handle the case when inputs have commas
3. User can directly paste a comma separated list of mutations (spaces will be trimmed). The list gets transformed into valid tags immediately, compliant to the rules and logic defined in the [suggestOptions ](https://github.com/cevo-public/cov-spectrum-website/blob/601be84331101ef9d1f497fa1b41ddca7e332be5/src/components/VariantSearch/VariantSearch.tsx#L108-L125) method.
4. User can also create valid tag by entering a mutation name then comma one by one (it behaves like the enter key) 

![copy paste comma separated list](https://user-images.githubusercontent.com/3387698/127647976-e537d573-b890-4c57-95c6-4e9af232bf8a.gif)

The following gif shows when there are invalid entries from the list. The valid mutation queries will be transformed into tags and the invalid queries text will remain in the input field and the drop down option "No Options" will be shown.  (I think this could be better and cleaner than showing an error message for those invalid queries.)

![copy paste comma separated list with invalid queries](https://user-images.githubusercontent.com/3387698/127648070-89e826d1-4955-4ffc-a0a1-bf55b4fd4bbf.gif)


The search on enter feature still works.


@TKGZ @chaoran-chen Please review the code and test the features. Thanks!

ningxie1991 committed:
Added search by comma separated list feature

1. Modified the AsyncSelect component by manually setting value, inputValue and menuIsOpen
2. Added handleInputChange, handleCommaSeparatedInput methods to handle the case when inputs have commas
3. User can either create valid tag by entering a mutation then comma one by one (it behaves like the enter key) or user can paste a comma separated list of mutations
4. The list gets transformed into valid tags immediately, compliant to the rules and logic defined in the suggestOptions method.

ningxie1991 committed:
Update VariantSearch.tsx

ningxie1991 committed:
Update VariantSearch.tsx

Fix prettier format check

chaoran-chen reviewed (changes requested):
Hi @ningxie1991, I only took a very brief look at the preview and it looks really good!

I just found one issue: it shouldn't be possible to insert two pangolin lineages (e.g., `B.1.1.7,B.1.617`). Maybe only accept the first one?

![image](https://user-images.githubusercontent.com/18666552/127653123-7d97b951-a449-4c08-9146-e591dbba71b6.png)

ningxie1991 commented:
> Hi @ningxie1991, I only took a very brief look at the preview and it looks really good!
> 
> I just found one issue: it shouldn't be possible to insert two pangolin lineages (e.g., `B.1.1.7,B.1.617`). Maybe only accept the first one?
> 
> ![image](https://user-images.githubusercontent.com/18666552/127653123-7d97b951-a449-4c08-9146-e591dbba71b6.png)

Good catch! When user enters one query then comma one by one this won't happen as the code checks for the max one pangolin lineage rule and won't allow user to enter more than one. But when user just pastes a pre-written comma separated list with more than one pangolin lineages, this happens. I will check and fix that.

ningxie1991 committed:
Update VariantSearch.tsx

Fix the bug that we should only allow max one pangolin lineage in the selected options while parsing the comma separated list.

ningxie1991 commented:
> > Hi @ningxie1991, I only took a very brief look at the preview and it looks really good!
> > I just found one issue: it shouldn't be possible to insert two pangolin lineages (e.g., `B.1.1.7,B.1.617`). Maybe only accept the first one?
> > ![image](https://user-images.githubusercontent.com/18666552/127653123-7d97b951-a449-4c08-9146-e591dbba71b6.png)
> 
> Good catch! When user enters one query then comma one by one this won't happen as the code checks for the max one pangolin lineage rule and won't allow user to enter more than one. But when user just pastes a pre-written comma separated list with more than one pangolin lineages, this happens. I will check and fix that.

This is fixed now. Only the first pangolin lineage will be accepted. @chaoran-chen

ningxie1991 committed:
Update VariantSearch.tsx

Added comment for clarification

chaoran-chen reviewed (approved):
Looks good to me! (Please just don't forget to fix the formatting)

spacephoton commented on src/components/VariantSearch/VariantSearch.tsx:
```diff
@@ -124,50 +130,146 @@ export const VariantSearch = ({ onVariantSelect }: Props) => {
+  const promiseOptions = (inputValue : string) => {
+    // resets the options to default (where input value is '') when menu is closed
+    if(!menuIsOpen){
+      inputValue = ''
```
Why not `setInputValue('')`?

Should we use `undefined` by default when no input value? (I don't know)

ningxie1991 replied:
`setInputValue('')` is called right after at line 205. But because the state change is async and doesn't happen immediately so I still passed the inputValue at line 208. This code is simulating the action that when an option is selected the option menu gets closed and the suggest options get reloaded (it does not happen automatically for selecting from using comma-separated list because onChange is not fired). 

https://github.com/cevo-public/cov-spectrum-website/blob/8944d23088d2d9ce0a803f4003ea52d05d54092e/src/components/VariantSearch/VariantSearch.tsx#L201-L208

The reason why I used `''` instead of `undefined` is because  line 262, the defaultOptions come from '' input. 

https://github.com/cevo-public/cov-spectrum-website/blob/8944d23088d2d9ce0a803f4003ea52d05d54092e/src/components/VariantSearch/VariantSearch.tsx#L262

spacephoton commented on src/components/VariantSearch/VariantSearch.tsx:
```diff
@@ -124,50 +130,146 @@ export const VariantSearch = ({ onVariantSelect }: Props) => {
+      query = query.trim()
+      const suggestions = suggestOptions(query);
+      const selectedOption = suggestions.find(option => option.value.toUpperCase() === query.toUpperCase());
+      // If query gets back valid a valid suggest option which is not yet selected,
```
Might be useful to have high level explanation comment on top of the function, rather than details inserted throughout. But this is just a styling thing, so if you prefer like this, no problem.

ningxie1991 replied:
Good suggestions! I added function doc and removed the inline comments.

spacephoton reviewed (approved):
Seems good!

To fix format just run the npm script
`npm run format`

ningxie1991 committed:
Update VariantSearch.tsx

Fix format and comments

ningxie1991 commented:
> Looks good to me! (Please just don't forget to fix the formatting)

Formatting is fixed with `npm run format`

ningxie1991 commented:
This is the one warning failing the `Check formatting`, which I am not sure if we can do anything about it.

![image](https://user-images.githubusercontent.com/3387698/128487113-64193925-4cc1-4fc5-a557-e70b8c95aed4.png)

ningxie1991 committed:
Update VariantSearch.tsx

Fix formatting issue

ningxie1991 committed:
Update VariantSearch.tsx

update format

ningxie1991 commented on src/components/VariantSearch/VariantSearch.tsx:
```diff
@@ -150,18 +256,28 @@ export const VariantSearch = ({ onVariantSelect }: Props) => {
+              setMenuIsOpen(false);
             } else if (change.action === 'remove-value' || change.action === 'pop-value') {
               setSelectedOptions(selectedOptions.filter(o => o.value !== change.removedValue.value));
+            } else if (change.action === 'clear') {
```
Fixed issue #216
